# Zomato Bangalore Restaurants Data Cleaning

**Importing Libraries**

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

**Reading Data**

In [2]:
df = pd.read_csv("zomato.csv")
df.head()

url  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2  https://www.zomato.com/SanchurroBangalore?cont...   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             address                   name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  online_order book_table   rate  votes                             phone  \
0          Yes        Yes  4.1/5    775    080 42297555\r\n+91 9743772233   
1          Yes         No  4.1/5    787                      080 41714161   
2          Yes         No  3.8/5    918                    +91 9663487993   
3           No         No  3.7/5     88                    +91 9620009302   
4           No         No  3.8/5    166  +91 8026612447\r\n+91 9901210005   

       location            rest_type  \
0  Banashankari        Casual Dining   
1  Banashankari        Casual Dining   
2  Banashankari  Cafe, Casual Dining   
3  Banashankari          Quick Bites   
4  Basavanagudi        Casual Dining   

                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines approx_cost(for two people)  \
0  North Indian, Mughlai, Chinese                         800   
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        reviews_list menu_item  \
0  [('Rated 4.0', 'RATED\n  A beautiful place to ...        []   
1  [('Rated 4.0', 'RATED\n  Had been here for din...        []   
2  [('Rated 3.0', "RATED\n  Ambience is not that ...        []   
3  [('Rated 4.0', "RATED\n  Great food and proper...        []   
4  [('Rated 4.0', 'RATED\n  Very good restaurant ...        []   

  listed_in(type) listed_in(city)  
0          Buffet    Banashankari  
1          Buffet    Banashankari  
2          Buffet    Banashankari  
3          Buffet    Banashankari  
4          Buffet    Banashankari

In [3]:
df.tail()

url  \
51712  https://www.zomato.com/bangalore/best-brews-fo...   
51713  https://www.zomato.com/bangalore/vinod-bar-and...   
51714  https://www.zomato.com/bangalore/plunge-sherat...   
51715  https://www.zomato.com/bangalore/chime-sherato...   
51716  https://www.zomato.com/bangalore/the-nest-the-...   

                                                 address  \
51712  Four Points by Sheraton Bengaluru, 43/3, White...   
51713  Number 10, Garudachar Palya, Mahadevapura, Whi...   
51714  Sheraton Grand Bengaluru Whitefield Hotel & Co...   
51715  Sheraton Grand Bengaluru Whitefield Hotel & Co...   
51716  ITPL Main Road, KIADB Export Promotion Industr...   

                                                    name online_order  \
51712  Best Brews - Four Points by Sheraton Bengaluru...           No   
51713                           Vinod Bar And Restaurant           No   
51714  Plunge - Sheraton Grand Bengaluru Whitefield H...           No   
51715  Chime - Sheraton Grand Bengaluru Whitefield Ho...           No   
51716                       The Nest - The Den Bengaluru           No   

      book_table    rate  votes           phone                    location  \
51712         No  3.6 /5     27    080 40301477                  Whitefield   
51713         No     NaN      0  +91 8197675843                  Whitefield   
51714         No     NaN      0             NaN                  Whitefield   
51715        Yes  4.3 /5    236    080 49652769  ITPL Main Road, Whitefield   
51716         No  3.4 /5     13  +91 8071117272  ITPL Main Road, Whitefield   

                rest_type                    dish_liked  \
51712                 Bar                           NaN   
51713                 Bar                           NaN   
51714                 Bar                           NaN   
51715                 Bar  Cocktails, Pizza, Buttermilk   
51716  Bar, Casual Dining                           NaN   

                                     cuisines approx_cost(for two people)  \
51712                             Continental                       1,500   
51713                             Finger Food                         600   
51714                             Finger Food                       2,000   
51715                             Finger Food                       2,500   
51716  Finger Food, North Indian, Continental                       1,500   

                                            reviews_list menu_item  \
51712  [('Rated 5.0', "RATED\n  Food and service are ...        []   
51713                                                 []        []   
51714                                                 []        []   
51715  [('Rated 4.0', 'RATED\n  Nice and friendly pla...        []   
51716  [('Rated 5.0', 'RATED\n  Great ambience , look...        []   

      listed_in(type) listed_in(city)  
51712   Pubs and bars      Whitefield  
51713   Pubs and bars      Whitefield  
51714   Pubs and bars      Whitefield  
51715   Pubs and bars      Whitefield  
51716   Pubs and bars      Whitefield

## **Data Cleaning**

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 17 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   url                          51717 non-null  object
 1   address                      51717 non-null  object
 2   name                         51717 non-null  object
 3   online_order                 51717 non-null  object
 4   book_table                   51717 non-null  object
 5   rate                         43942 non-null  object
 6   votes                        51717 non-null  int64 
 7   phone                        50509 non-null  object
 8   location                     51696 non-null  object
 9   rest_type                    51490 non-null  object
 10  dish_liked                   23639 non-null  object
 11  cuisines                     51672 non-null  object
 12  approx_cost(for two people)  51371 non-null  object
 13  reviews_list                 51

In [5]:
df.nunique()

url                            51717
address                        11495
name                            8792
online_order                       2
book_table                         2
rate                              64
votes                           2328
phone                          14926
location                          93
rest_type                         93
dish_liked                      5271
cuisines                        2723
approx_cost(for two people)       70
reviews_list                   22513
menu_item                       9098
listed_in(type)                    7
listed_in(city)                   30
dtype: int64

### **Fixing url column values**

In [6]:
df.url.values[:5]

array(['https://www.zomato.com/bangalore/jalsa-banashankari?context=eyJzZSI6eyJlIjpbNTg2OTQsIjE4Mzc1NDc0IiwiNTkwOTAiLCIxODM4Mjk0NCIsIjE4MjI0Njc2IiwiNTkyODkiLCIxODM3MzM4NiJdLCJ0IjoiUmVzdGF1cmFudHMgaW4gQmFuYXNoYW5rYXJpIHNlcnZpbmcgQnVmZmV0In19',
       'https://www.zomato.com/bangalore/spice-elephant-banashankari?context=eyJzZSI6eyJlIjpbIjU4Njk0IiwxODM3NTQ3NCwiNTkwOTAiLCIxODM4Mjk0NCIsIjE4MjI0Njc2IiwiNTkyODkiLCIxODM3MzM4NiJdLCJ0IjoiUmVzdGF1cmFudHMgaW4gQmFuYXNoYW5rYXJpIHNlcnZpbmcgQnVmZmV0In19',
       'https://www.zomato.com/SanchurroBangalore?context=eyJzZSI6eyJlIjpbIjU4Njk0IiwiMTgzNzU0NzQiLDU5MDkwLCIxODM4Mjk0NCIsIjE4MjI0Njc2IiwiNTkyODkiLCIxODM3MzM4NiJdLCJ0IjoiUmVzdGF1cmFudHMgaW4gQmFuYXNoYW5rYXJpIHNlcnZpbmcgQnVmZmV0In19',
       'https://www.zomato.com/bangalore/addhuri-udupi-bhojana-banashankari?context=eyJzZSI6eyJlIjpbIjU4Njk0IiwiMTgzNzU0NzQiLCI1OTA5MCIsMTgzODI5NDQsIjE4MjI0Njc2IiwiNTkyODkiLCIxODM3MzM4NiJdLCJ0IjoiUmVzdGF1cmFudHMgaW4gQmFuYXNoYW5rYXJpIHNlcnZpbmcgQnVmZmV0In19',
       'https

In [7]:
df['url'] = df.url.str.split('?',expand=True)[0]

In [8]:
df.url.values[:5]

array(['https://www.zomato.com/bangalore/jalsa-banashankari',
       'https://www.zomato.com/bangalore/spice-elephant-banashankari',
       'https://www.zomato.com/SanchurroBangalore',
       'https://www.zomato.com/bangalore/addhuri-udupi-bhojana-banashankari',
       'https://www.zomato.com/bangalore/grand-village-basavanagudi'],
      dtype=object)

In [9]:
df.url.nunique()

12453

*Many Restaurants are occuring in more than one row*

### **name**

In [10]:
df.name.value_counts()

name
Cafe Coffee Day                                            96
Onesta                                                     85
Just Bake                                                  73
Empire Restaurant                                          71
Five Star Chicken                                          70
                                                           ..
Natis                                                       1
Shreyas                                                     1
Vishwas Kabab Centre                                        1
Taj Biryani Centre                                          1
Plunge - Sheraton Grand Bengaluru Whitefield Hotel &...     1
Name: count, Length: 8792, dtype: int64

*Many of the Restaurant has multiple rows*

### **Phone**

**Dropping Phone Column as it is not needed for the data analysis** 

In [11]:
df.drop("phone",axis=1,inplace=True)

### **menu_list and reviews_list**

**Transforming empty list values in menu_list and reviews_list columns to null**

In [12]:
df["menu_item"] = df.menu_item.replace("[]",np.nan)
df["reviews_list"] = df.reviews_list.replace("[]",np.nan)

In [13]:
df.isnull().sum()

url                                0
address                            0
name                               0
online_order                       0
book_table                         0
rate                            7775
votes                              0
location                          21
rest_type                        227
dish_liked                     28078
cuisines                          45
approx_cost(for two people)      346
reviews_list                    7595
menu_item                      39617
listed_in(type)                    0
listed_in(city)                    0
dtype: int64

**Dropping dish_liked and menu_item columns as they have more than 50% values as null values**

In [14]:
df.drop(["dish_liked","menu_item"],axis=1,inplace=True)

In [15]:
df.head()

url  \
0  https://www.zomato.com/bangalore/jalsa-banasha...   
1  https://www.zomato.com/bangalore/spice-elephan...   
2          https://www.zomato.com/SanchurroBangalore   
3  https://www.zomato.com/bangalore/addhuri-udupi...   
4  https://www.zomato.com/bangalore/grand-village...   

                                             address                   name  \
0  942, 21st Main Road, 2nd Stage, Banashankari, ...                  Jalsa   
1  2nd Floor, 80 Feet Road, Near Big Bazaar, 6th ...         Spice Elephant   
2  1112, Next to KIMS Medical College, 17th Cross...        San Churro Cafe   
3  1st Floor, Annakuteera, 3rd Stage, Banashankar...  Addhuri Udupi Bhojana   
4  10, 3rd Floor, Lakshmi Associates, Gandhi Baza...          Grand Village   

  online_order book_table   rate  votes      location            rest_type  \
0          Yes        Yes  4.1/5    775  Banashankari        Casual Dining   
1          Yes         No  4.1/5    787  Banashankari        Casual Dining   
2          Yes         No  3.8/5    918  Banashankari  Cafe, Casual Dining   
3           No         No  3.7/5     88  Banashankari          Quick Bites   
4           No         No  3.8/5    166  Basavanagudi        Casual Dining   

                         cuisines approx_cost(for two people)  \
0  North Indian, Mughlai, Chinese                         800   
1     Chinese, North Indian, Thai                         800   
2          Cafe, Mexican, Italian                         800   
3      South Indian, North Indian                         300   
4        North Indian, Rajasthani                         600   

                                        reviews_list listed_in(type)  \
0  [('Rated 4.0', 'RATED\n  A beautiful place to ...          Buffet   
1  [('Rated 4.0', 'RATED\n  Had been here for din...          Buffet   
2  [('Rated 3.0', "RATED\n  Ambience is not that ...          Buffet   
3  [('Rated 4.0', "RATED\n  Great food and proper...          Buffet   
4  [('Rated 4.0', 'RATED\n  Very good restaurant ...          Buffet   

  listed_in(city)  
0    Banashankari  
1    Banashankari  
2    Banashankari  
3    Banashankari  
4    Banashankari

### **rate**

**Removing '/5' in rate column and changing its datatype to float**

In [16]:
pd.set_option('display.max_rows', 100)
df.rate.value_counts(dropna=False)

rate
NaN       7775
NEW       2208
3.9/5     2098
3.8/5     2022
3.7/5     2011
3.9 /5    1874
3.8 /5    1851
3.7 /5    1810
3.6/5     1773
4.0/5     1609
4.0 /5    1574
3.6 /5    1543
4.1/5     1474
4.1 /5    1474
3.5/5     1431
3.5 /5    1353
3.4/5     1259
3.4 /5    1217
3.3/5     1168
4.2 /5    1165
3.3 /5    1142
4.2/5     1019
3.2/5     1006
4.3 /5     917
3.2 /5     867
3.1/5      862
4.3/5      776
3.1 /5     699
4.4 /5     628
3.0/5      558
4.4/5      519
3.0 /5     465
2.9/5      427
4.5 /5     409
2.9 /5     375
2.8/5      313
2.8 /5     287
4.5/5      247
4.6 /5     175
2.7/5      170
2.6/5      143
2.7 /5     137
4.6/5      125
2.6 /5     117
4.7 /5      86
4.7/5       81
-           69
2.5 /5      56
2.5/5       45
4.8 /5      43
2.4/5       40
4.9 /5      30
2.4 /5      30
2.3/5       28
4.9/5       25
2.3 /5      23
4.8/5       23
2.2/5       19
2.1 /5      13
2.1/5       11
2.2 /5       7
2.0 /5       7
2.0/5        4
1.8 /5       3
1.8/5        2
Name: count, dtype: 

*Changing the "NEW" value in rate column to 0 and "-" to null*

In [17]:
df["rate"] = df.rate.replace({"NEW":"0.0","-":np.nan})
df["rate"] = df.rate.str.replace('/5','')
df["rate"]

0         4.1
1         4.1
2         3.8
3         3.7
4         3.8
         ... 
51712    3.6 
51713     NaN
51714     NaN
51715    4.3 
51716    3.4 
Name: rate, Length: 51717, dtype: object

In [18]:
df["rate"] = df.rate.astype('float')
df.rename({'rate':'rating'},axis=1,inplace=True)

### **cost(for two people)**

**Changing approx_cost(for two people) datatype to float**

In [19]:
df.rename({'approx_cost(for two people)':'cost_for_two'},axis=1,inplace=True)

In [20]:
df.cost_for_two.value_counts()

cost_for_two
300      7576
400      6562
500      4980
200      4857
600      3714
250      2959
800      2285
150      2066
700      1948
350      1763
1,000    1637
450      1417
1,200     993
100       993
1,500     971
650       776
550       761
750       758
900       700
1,300     516
1,100     512
1,400     473
2,000     363
1,600     266
1,700     247
1,800     203
850       166
3,000     162
2,500     146
2,200      78
1,900      70
2,100      67
950        62
2,800      45
4,000      29
3,500      25
120        24
2,400      23
180        20
1,350      18
3,400      13
2,300      11
2,600      10
80         10
230        10
1,250       9
40          8
50          8
130         8
1,650       6
1,450       5
199         4
330         4
4,100       4
1,050       4
2,700       3
70          3
6,000       2
4,500       2
190         2
360         2
240         2
3,200       2
140         2
560         1
60          1
5,000       1
3,700       1
469         1
160         1
Name: c

*cost_for_two values have "," in values above 999\
Removing "," in values and then changing dtype to float*

In [21]:
df["cost_for_two"] = df.cost_for_two.str.replace(",","").astype('float')
df["cost_for_two"]

0         800.0
1         800.0
2         800.0
3         300.0
4         600.0
          ...  
51712    1500.0
51713     600.0
51714    2000.0
51715    2500.0
51716    1500.0
Name: cost_for_two, Length: 51717, dtype: float64

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   url              51717 non-null  object 
 1   address          51717 non-null  object 
 2   name             51717 non-null  object 
 3   online_order     51717 non-null  object 
 4   book_table       51717 non-null  object 
 5   rating           43873 non-null  float64
 6   votes            51717 non-null  int64  
 7   location         51696 non-null  object 
 8   rest_type        51490 non-null  object 
 9   cuisines         51672 non-null  object 
 10  cost_for_two     51371 non-null  float64
 11  reviews_list     44122 non-null  object 
 12  listed_in(type)  51717 non-null  object 
 13  listed_in(city)  51717 non-null  object 
dtypes: float64(2), int64(1), object(11)
memory usage: 5.5+ MB


### **rest_type**

In [23]:
df.rest_type.value_counts()

rest_type
Quick Bites                      19132
Casual Dining                    10330
Cafe                              3732
Delivery                          2604
Dessert Parlor                    2263
Takeaway, Delivery                2037
Casual Dining, Bar                1154
Bakery                            1141
Beverage Shop                      867
Bar                                697
Food Court                         624
Sweet Shop                         468
Bar, Casual Dining                 425
Lounge                             396
Pub                                357
Fine Dining                        346
Casual Dining, Cafe                319
Beverage Shop, Quick Bites         298
Bakery, Quick Bites                289
Mess                               267
Pub, Casual Dining                 255
Sweet Shop, Quick Bites            178
Kiosk                              176
Dessert Parlor, Cafe               175
Cafe, Casual Dining                173
Cafe, Bakery   

In [24]:
df.rest_type.isnull().sum()

227

In [25]:
df.rest_type.str.split(", ",expand=True)

0              1
0      Casual Dining           None
1      Casual Dining           None
2               Cafe  Casual Dining
3        Quick Bites           None
4      Casual Dining           None
...              ...            ...
51712            Bar           None
51713            Bar           None
51714            Bar           None
51715            Bar           None
51716            Bar  Casual Dining

[51717 rows x 2 columns]

In [26]:
df['rest_type'].str.split(', ').explode().value_counts(dropna=False).head(20)

rest_type
Quick Bites       20639
Casual Dining     13057
Cafe               5074
Delivery           4641
Dessert Parlor     3211
Bar                2457
Takeaway           2157
Bakery             1998
Beverage Shop      1471
Pub                 950
Food Court          765
Sweet Shop          745
Lounge              555
Fine Dining         405
Microbrewery        390
Mess                274
NaN                 227
Kiosk               192
Food Truck           84
Confectionery        47
Name: count, dtype: int64

**Filling Null values in rest_type corresponding to each listed_in(type)**

In [27]:
from statistics import mode

In [28]:
lType_to_restType = df[["url","rest_type","listed_in(type)"]].drop_duplicates().groupby("listed_in(type)").agg(
    mode_rest_type=('rest_type', mode),
    mode_count=('rest_type', lambda x: x.tolist().count(mode(x)))
)
lType_to_restType

mode_rest_type  mode_count
listed_in(type)                                   
Buffet                   Casual Dining         237
Cafes                             Cafe         503
Delivery                   Quick Bites        3712
Desserts                Dessert Parlor         517
Dine-out                   Quick Bites        4274
Drinks & nightlife                 Bar          80
Pubs and bars       Casual Dining, Bar         107

In [29]:
new_rest_type = df.set_index(df["listed_in(type)"])["rest_type"].fillna(lType_to_restType["mode_rest_type"])
new_rest_type

listed_in(type)
Buffet                 Casual Dining
Buffet                 Casual Dining
Buffet           Cafe, Casual Dining
Buffet                   Quick Bites
Buffet                 Casual Dining
                        ...         
Pubs and bars                    Bar
Pubs and bars                    Bar
Pubs and bars                    Bar
Pubs and bars                    Bar
Pubs and bars     Bar, Casual Dining
Name: rest_type, Length: 51717, dtype: object

In [30]:
df["rest_type"] = new_rest_type.reset_index(drop=True)

In [31]:
df.rest_type.isnull().sum()

0

### **cuisines**

In [32]:
df.cuisines.value_counts()

cuisines
North Indian                                   2913
North Indian, Chinese                          2385
South Indian                                   1828
Biryani                                         918
Bakery, Desserts                                911
                                               ... 
North Indian, Chinese, South Indian, Juices       1
North Indian, Chinese, Kebab, Mughlai             1
Chinese, Vietnamese, Thai, Malaysian              1
Arabian, Lebanese, Chinese, Rolls                 1
North Indian, Chinese, Arabian, Momos             1
Name: count, Length: 2723, dtype: int64

In [33]:
df.cuisines.isnull().sum()

45

In [34]:
df["cuisines"] = df.cuisines.fillna("Unknown")

In [35]:
df.cuisines.isnull().sum()

0

In [36]:
df.cuisines.str.split(", ",expand=True).dropna().drop_duplicates()

0            1                 2               3  \
55     Healthy Food      Chinese           Biryani    North Indian   
195             BBQ      Arabian             Rolls         Chinese   
864         Arabian      Biryani           Chinese  Middle Eastern   
900        American       Bakery         Beverages            Cafe   
944         Biryani       Kerala           Mughlai     Street Food   
1112        Seafood      Biryani      South Indian    North Indian   
2508           Cafe  Continental         Beverages    Healthy Food   
2526   South Indian      Chinese         Beverages            Cafe   
3451    Continental      Biryani          Desserts         Italian   
3895       American      Italian            Bakery        Desserts   
4248         Andhra    Beverages           Biryani         Chinese   
4819    Continental         Thai         Malaysian         Burmese   
6573            BBQ      Biryani  Charcoal Chicken         Chinese   
6785       American         Cafe       Continental          French   
8267    Continental      Italian      North Indian           Asian   
8853          Salad      Biryani           Chinese    North Indian   
9697        Seafood      Biryani      South Indian       Chettinad   
13368   Continental      Mexican            Burger        Desserts   
13517   Street Food     American          Sandwich       Fast Food   
16391         Pizza       Burger          Sandwich          Coffee   
17561          Cafe  Continental         Beverages          Burger   
17599          Cafe     American             Salad     Continental   
17607         Salad         Cafe       Continental           Steak   
17872        Burger     American         Beverages        Desserts   
17908   Continental      Seafood           Biryani        Desserts   
17932   Continental       Burger          Desserts        Sandwich   
18785   Continental     American             Pizza    North Indian   
24245         Asian     European           Italian          Korean   
28002       Arabian          BBQ           Biryani         Chinese   
37800          Cafe  Continental         Fast Food           Pizza   
40525   Continental       Burger      South Indian        Sandwich   
40682   Continental      Biryani             Pizza    North Indian   
46181   Continental        Asian      Healthy Food          Burger   
49355       Bengali      Chinese         Fast Food           Kebab   

                  4             5              6             7  
55      Continental         Salad       American        Burger  
195    North Indian        Juices          Kebab      Desserts  
864       Fast Food  North Indian        Mughlai       Seafood  
900    Healthy Food        Juices   North Indian      Sandwich  
944    North Indian       Chinese   South Indian        Andhra  
1112        Chinese       Arabian          Kebab        Kerala  
2508       Desserts      Sandwich          Pizza         Salad  
2526         Coffee     Fast Food          Momos  North Indian  
3451   North Indian       Chinese  Modern Indian         Asian  
3895       Sandwich     Beverages          Salad         Rolls  
4248      Fast Food    Hyderabadi   North Indian  South Indian  
4819         Korean    Vietnamese        Chinese  North Indian  
6573   North Indian        Kerala        Seafood   Street Food  
6785         Burger       Mexican       Desserts         Pizza  
8267         Bakery       Chinese       European      Japanese  
8853         Kerala      Desserts        Seafood  South Indian  
9697        Chinese       Arabian         Andhra        Kerala  
13368       Italian         Pizza   North Indian       Chinese  
13517     Beverages        Burger       Desserts         Pizza  
16391      Desserts   Continental          Salad         Asian  
17561      Desserts         Pizza         Bakery      European  
17599  Healthy Food       Italian         Coffee      Desserts  
17607       Italian        Burger       Desserts         Pizza  
17872  

In [37]:
df['cuisines'].str.split(', ').explode().value_counts().head(20)

cuisines
North Indian    21085
Chinese         15547
South Indian     8644
Fast Food        8096
Biryani          6492
Continental      5765
Desserts         5633
Cafe             5303
Beverages        4747
Italian          3389
Bakery           2840
Street Food      2595
Pizza            2073
Burger           2009
Seafood          1810
Ice Cream        1774
Andhra           1760
Mughlai          1570
Rolls            1470
American         1461
Name: count, dtype: int64

### **reviews_list**

In [38]:
df.reviews_list.count()

44122

In [39]:
df.reviews_list.isnull().sum()

7595

In [40]:
df.reviews_list.value_counts()

reviews_list
[('Rated 5.0', "RATED\n  This lobby cafe offers beverages alongside a wide delicious line up of desserts ranging from apple pies to pastries to Indian sweets. It's set up in a warm ambiance. We visited during the Christmas season and the cafe has a host of edible gingerbread houses and festively themed cakes. The staff is courteous and prompt. We loved the waffles and doughnuts.\n\nalthingsbeautiful.wordpress.com")]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        

In [41]:
df.reviews_list.str.startswith("[").count()

44122

*Every Value in review list is in correct format*

In [42]:
df.reviews_list[0]

'[(\'Rated 4.0\', \'RATED\\n  A beautiful place to dine in.The interiors take you back to the Mughal era. The lightings are just perfect.We went there on the occasion of Christmas and so they had only limited items available. But the taste and service was not compromised at all.The only complaint is that the breads could have been better.Would surely like to come here again.\'), (\'Rated 4.0\', \'RATED\\n  I was here for dinner with my family on a weekday. The restaurant was completely empty. Ambience is good with some good old hindi music. Seating arrangement are good too. We ordered masala papad, panner and baby corn starters, lemon and corrionder soup, butter roti, olive and chilli paratha. Food was fresh and good, service is good too. Good for family hangout.\\nCheers\'), (\'Rated 2.0\', \'RATED\\n  Its a restaurant near to Banashankari BDA. Me along with few of my office friends visited to have buffet but unfortunately they only provide veg buffet. On inquiring they said this plac

*Extracting Rating Values from reviews list and taking their mean to fill null values in rating column*

In [43]:
df.reviews_list[0:1].str.findall(r'Rated (\d+.\d)').values

array([list(['4.0', '4.0', '2.0', '4.0', '5.0', '5.0', '4.0', '4.0', '5.0', '4.0', '4.0', '4.0'])],
      dtype=object)

In [44]:
np.array(df.reviews_list[0:1].str.findall(r'Rated (\d+.\d)').values[0]).astype('float')

array([4., 4., 2., 4., 5., 5., 4., 4., 5., 4., 4., 4.])

In [45]:
np.array(df.reviews_list[0:1].str.findall(r'Rated (\d+.\d)').values[0]).astype('float').mean()

4.083333333333333

In [46]:
df.rating[0]

4.1

*Quite similar to the actual rating value*

In [47]:
df.reviews_list[0:5].str.findall(r'Rated (\d+.\d)')

0    [4.0, 4.0, 2.0, 4.0, 5.0, 5.0, 4.0, 4.0, 5.0, ...
1    [4.0, 3.0, 3.0, 4.0, 4.0, 5.0, 4.0, 2.0, 4.0, ...
2    [3.0, 3.0, 4.0, 3.0, 3.0, 1.0, 3.0, 1.0, 4.0, ...
3    [4.0, 2.0, 4.0, 2.0, 4.0, 5.0, 4.0, 5.0, 3.5, ...
4                                           [4.0, 4.0]
Name: reviews_list, dtype: object

In [48]:
avg_rating = df['reviews_list'].str.findall(r'Rated (\d+\.\d)').apply(
    lambda ratings: round(np.array(ratings, dtype='float').mean(),1) if ratings else None
)

In [49]:
avg_rating

0        4.1
1        3.6
2        3.2
3        3.7
4        4.0
        ... 
51712    4.1
51713    NaN
51714    NaN
51715    4.0
51716    5.0
Name: reviews_list, Length: 51717, dtype: float64

In [50]:
df['rating'].fillna(avg_rating,inplace=True)

In [51]:
df.rating.isnull().sum()

5319

In [52]:
df.drop("reviews_list",axis=1,inplace=True)

### **votes**

In [53]:
df.votes.value_counts()

votes
0       10027
4        1140
6         992
7         872
9         738
        ...  
3673        1
1862        1
3909        1
2155        1
843         1
Name: count, Length: 2328, dtype: int64

In [54]:
df.votes.dtype

dtype('int64')

### **listed_in(type)**

In [55]:
df["listed_in(type)"].value_counts()

listed_in(type)
Delivery              25942
Dine-out              17779
Desserts               3593
Cafes                  1723
Drinks & nightlife     1101
Buffet                  882
Pubs and bars           697
Name: count, dtype: int64

*Grouping all listed in type values for one url into one row*

In [56]:
df.groupby("url",as_index=False)["listed_in(type)"].apply(lambda x :', '.join(x.unique()))

url  \
0      https://www.zomato.com/24-7RestaurantBangalore   
1               https://www.zomato.com/29degreesnorth   
2               https://www.zomato.com/ArborBrewIndia   
3             https://www.zomato.com/BaluchiBangalore   
4                    https://www.zomato.com/Barebones   
...                                               ...   
12448          https://www.zomato.com/thesugarfactory   
12449          https://www.zomato.com/theunderpassblr   
12450               https://www.zomato.com/ucfastfood   
12451             https://www.zomato.com/umashomefood   
12452        https://www.zomato.com/zyksha-food-truck   

                                   listed_in(type)  
0                                 Buffet, Dine-out  
1                               Delivery, Dine-out  
2      Dine-out, Drinks & nightlife, Pubs and bars  
3                                         Dine-out  
4      Dine-out, Drinks & nightlife, Pubs and bars  
...                                            ...  
12448                 Dine-out, Drinks & nightlife  
12449                 Dine-out, Drinks & nightlife  
12450                           Delivery, Dine-out  
12451                                     Delivery  
12452                           Delivery, Dine-out  

[12453 rows x 2 columns]

In [57]:
df[df.name=="Jalsa"]

url  \
0      https://www.zomato.com/bangalore/jalsa-banasha...   
456    https://www.zomato.com/bangalore/jalsa-banasha...   
559    https://www.zomato.com/bangalore/jalsa-banasha...   
2488   https://www.zomato.com/bangalore/jalsa-banasha...   
2966   https://www.zomato.com/bangalore/jalsa-banasha...   
3462   https://www.zomato.com/bangalore/jalsa-banasha...   
7607   https://www.zomato.com/bangalore/jalsa-maratha...   
19401  https://www.zomato.com/bangalore/jalsa-banasha...   
20399  https://www.zomato.com/bangalore/jalsa-banasha...   
21302  https://www.zomato.com/bangalore/jalsa-banasha...   
41364  https://www.zomato.com/bangalore/jalsa-maratha...   

                                                 address   name online_order  \
0      942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   
456    942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   
559    942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   
2488   942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   
2966   942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   
3462   942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   
7607   25/7, Doddannakundi, Outer Ring Road, Mahadeva...  Jalsa          Yes   
19401  942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   
20399  942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   
21302  942, 21st Main Road, 2nd Stage, Banashankari, ...  Jalsa          Yes   
41364  25/7, Doddannakundi, Outer Ring Road, Mahadeva...  Jalsa          Yes   

      book_table  rating  votes      location      rest_type  \
0            Yes     4.1    775  Banashankari  Casual Dining   
456          Yes     4.1    775  Banashankari  Casual Dining   
559          Yes     4.1    775  Banashankari  Casual Dining   
2488         Yes     4.1    775  Banashankari  Casual Dining   
2966         Yes     4.1    775  Banashankari  Casual Dining   
3462         Yes     4.1    804  Banashankari  Casual Dining   
7607         Yes     4.2   2173  Marathahalli  Casual Dining   
19401        Yes     4.1    783  Banashankari  Casual Dining   
20399        Yes     4.1    783  Banashankari  Casual Dining   
21302        Yes     4.1    783  Banashankari  Casual Dining   
41364        Yes     4.2   2189  Marathahalli  Casual Dining   

                             cuisines  cost_for_two listed_in(type)  \
0      North Indian, Mughlai, Chinese         800.0          Buffet   
456    North Indian, Mughlai, Chinese         800.0        Delivery   
559    North Indian, Mughlai, Chinese         800.0        Dine-out   
2488   North Indian, Mughlai, Chinese         800.0          Buffet   
2966   North Indian, Mughlai, Chinese         800.0        Delivery   
3462   North Indian, Mughlai, Chinese         800.0        Dine-out   
7607            North Indian, Mughlai        1500.0        Delivery   
19401  North Indian, Mughlai, Chinese         800.0          Buffet   
20399  North Indian, Mughlai, Chinese         800.0        Delivery   
21302  North Indian, Mughlai, Chinese         800.0        Dine-out   
41364           North Indian, Mughlai        1500.0        Delivery   

      listed_in(city)  
0        Banashankari  
456      Banashankari  
559      Banashankari  
2488     Basavanagudi  
2966     Basavanagudi  
3462     Basavanagudi  
7607      Brookefield  
19401       Jayanagar  
20399       Jayanagar  
21302       Jayanagar  
41364    Marathahalli

### **location**

In [58]:
df["location"].value_counts()

location
BTM                              5124
HSR                              2523
Koramangala 5th Block            2504
JP Nagar                         2235
Whitefield                       2144
Indiranagar                      2083
Jayanagar                        1926
Marathahalli                     1846
Bannerghatta Road                1630
Bellandur                        1286
Electronic City                  1258
Koramangala 1st Block            1238
Brigade Road                     1218
Koramangala 7th Block            1181
Koramangala 6th Block            1156
Sarjapur Road                    1065
Ulsoor                           1023
Koramangala 4th Block            1017
MG Road                           918
Banashankari                      906
Kalyan Nagar                      853
Richmond Road                     812
Frazer Town                       727
Malleshwaram                      725
Basavanagudi                      684
Residency Road                    675
Ban

In [59]:
df.location.isnull().sum()

21

In [60]:
df[df.address.notnull() & df.location.isnull()].shape

(21, 13)

*So we can fill location values from address*

In [62]:
locFromAddr = df.address.str.split(', ').str[-2]
locFromAddr

0          Banashankari
1          Banashankari
2          Banashankari
3          Banashankari
4          Basavanagudi
              ...      
51712        Whitefield
51713        Whitefield
51714        Whitefield
51715    ITPL Main Road
51716    ITPL Main Road
Name: address, Length: 51717, dtype: object

In [63]:
locFromAddr[df.location.isnull()]

1662              Bommanahalli
9874              Bommanahalli
13693          Electronic City
16351             Bommanahalli
26432             Kalyan Nagar
26519    Koramangala 8th Block
27672    Koramangala 6th Block
28611    Koramangala 6th Block
29298    Koramangala 8th Block
30604    Koramangala 6th Block
31400    Koramangala 6th Block
32131    Koramangala 8th Block
33457    Koramangala 4th Block
34755    Koramangala 8th Block
40354              Rajajinagar
40556             Marathahalli
41980             Marathahalli
46586             Cholourpalya
46609        Govindaraja Nagar
46752              Rajajinagar
47134             Cholourpalya
Name: address, dtype: object

In [64]:
df.location.fillna(locFromAddr).isnull().sum()

0

In [65]:
df["location"] = df.location.fillna(locFromAddr)

In [66]:
df.drop("address",axis=1,inplace=True)

### **listed_in(city)**

In [67]:
df["listed_in(city)"].value_counts()

listed_in(city)
BTM                      3279
Koramangala 7th Block    2938
Koramangala 5th Block    2836
Koramangala 4th Block    2779
Koramangala 6th Block    2623
Jayanagar                2371
JP Nagar                 2096
Indiranagar              1860
Church Street            1827
MG Road                  1811
Brigade Road             1769
Lavelle Road             1744
HSR                      1741
Marathahalli             1659
Residency Road           1620
Whitefield               1620
Bannerghatta Road        1617
Brookefield              1518
Old Airport Road         1425
Kammanahalli             1329
Kalyan Nagar             1309
Basavanagudi             1266
Sarjapur Road            1261
Electronic City          1229
Bellandur                1227
Frazer Town              1185
Malleshwaram             1096
Rajajinagar              1079
Banashankari              863
New BEL Road              740
Name: count, dtype: int64

*Not much significance*

In [68]:
df.drop("listed_in(city)",axis=1,inplace=True)

### **Grouping multiple rows corresponding to the same restaurant into a single consolidated row**
### **by applying different aggregation functions to each column based on the type of data**

In [69]:
df.groupby("url",as_index=False).apply(lambda df :df).head(6)

url  \
0 39419  https://www.zomato.com/24-7RestaurantBangalore   
  40269  https://www.zomato.com/24-7RestaurantBangalore   
1 24190           https://www.zomato.com/29degreesnorth   
  25039           https://www.zomato.com/29degreesnorth   
  25543           https://www.zomato.com/29degreesnorth   
  26315           https://www.zomato.com/29degreesnorth   

                                     name online_order book_table  rating  \
0 39419  24/7 - The Lalit Ashok Bangalore           No        Yes     4.0   
  40269  24/7 - The Lalit Ashok Bangalore           No        Yes     4.0   
1 24190                  29 Degrees North          Yes         No     3.9   
  25039                  29 Degrees North          Yes         No     3.9   
  25543                  29 Degrees North          Yes         No     3.9   
  26315                  29 Degrees North          Yes         No     3.9   

         votes       location    rest_type  \
0 39419    344  Seshadripuram  Fine Dining   
  40269    344  Seshadripuram  Fine Dining   
1 24190    283      Banaswadi  Quick Bites   
  25039    283      Banaswadi  Quick Bites   
  25543    283      Banaswadi  Quick Bites   
  26315    287      Banaswadi  Quick Bites   

                                            cuisines  cost_for_two  \
0 39419  Continental, North Indian, Italian, Chinese        3000.0   
  40269  Continental, North Indian, Italian, Chinese        3000.0   
1 24190         North Indian, Fast Food, Street Food         250.0   
  25039         North Indian, Fast Food, Street Food         250.0   
  25543         North Indian, Fast Food, Street Food         250.0   
  26315         North Indian, Fast Food, Street Food         250.0   

        listed_in(type)  
0 39419          Buffet  
  40269        Dine-out  
1 24190        Delivery  
  25039        Dine-out  
  25543        Delivery  
  26315        Dine-out

In [70]:
df.groupby(['url', 'name'])['votes'].nunique().reset_index().query('votes > 1')

url  \
1            https://www.zomato.com/29degreesnorth   
2            https://www.zomato.com/ArborBrewIndia   
6        https://www.zomato.com/ChurchStreetSocial   
7                  https://www.zomato.com/Cilantro   
8      https://www.zomato.com/FreshMenuIndiranagar   
...                                            ...   
12478              https://www.zomato.com/taantraa   
12479        https://www.zomato.com/techieskitchen   
12484          https://www.zomato.com/thepurplepan   
12485       https://www.zomato.com/thesugarfactory   
12486       https://www.zomato.com/theunderpassblr   

                                  name  votes  
1                     29 Degrees North      2  
2                Arbor Brewing Company      7  
6                 Church Street Social      7  
7                             Cilantro      3  
8                            FreshMenu      2  
...                                ...    ...  
12478    Taantraa - Organic Handbaking      2  
12479                  Techies Kitchen      2  
12484                   The Purple Pan      2  
12485  The Sugar Factory - Le Meridien      3  
12486         The Underpass Pub & Grub      3  

[3931 rows x 3 columns]

*Many restaurants has different votes values for different listed_type*\
*We will take only maximum of vote values for each restaurant*

In [71]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   url              51717 non-null  object 
 1   name             51717 non-null  object 
 2   online_order     51717 non-null  object 
 3   book_table       51717 non-null  object 
 4   rating           46398 non-null  float64
 5   votes            51717 non-null  int64  
 6   location         51717 non-null  object 
 7   rest_type        51717 non-null  object 
 8   cuisines         51717 non-null  object 
 9   cost_for_two     51371 non-null  float64
 10  listed_in(type)  51717 non-null  object 
dtypes: float64(2), int64(1), object(8)
memory usage: 4.3+ MB


In [72]:
df.url.nunique()

12453

*We want to have only 12453 rows representing each unique restaurant*

In [73]:
df.isnull().sum()

url                   0
name                  0
online_order          0
book_table            0
rating             5319
votes                 0
location              0
rest_type             0
cuisines              0
cost_for_two        346
listed_in(type)       0
dtype: int64

*We will fill these null values after grouping*

In [74]:
newDf = df.groupby("url",as_index=False).agg({
    "name" : "first",
    "online_order" : mode,
    "book_table" : mode,
    "rating" : "mean",
    "votes" : "max",
    "location" : mode,
    "rest_type" : mode,
    "cuisines" : mode,
    "cost_for_two" : "mean",
    "listed_in(type)" : lambda x : ", ".join(sorted(set(x))) 
})

In [75]:
newDf.head()

url  \
0  https://www.zomato.com/24-7RestaurantBangalore   
1           https://www.zomato.com/29degreesnorth   
2           https://www.zomato.com/ArborBrewIndia   
3         https://www.zomato.com/BaluchiBangalore   
4                https://www.zomato.com/Barebones   

                                  name online_order book_table  rating  votes  \
0     24/7 - The Lalit Ashok Bangalore           No        Yes     4.0    344   
1                     29 Degrees North          Yes         No     3.9    287   
2                Arbor Brewing Company           No        Yes     4.5   8419   
3  Baluchi - The Lalit Ashok Bangalore           No        Yes     4.0    134   
4          Barebones - The Balcony Bar           No        Yes     4.2   2445   

        location           rest_type  \
0  Seshadripuram         Fine Dining   
1      Banaswadi         Quick Bites   
2   Brigade Road   Pub, Microbrewery   
3  Seshadripuram         Fine Dining   
4    Indiranagar  Bar, Casual Dining   

                                      cuisines  cost_for_two  \
0  Continental, North Indian, Italian, Chinese        3000.0   
1         North Indian, Fast Food, Street Food         250.0   
2                        American, Continental        2000.0   
3              North Indian, Kashmiri, Mughlai        3700.0   
4   Chinese, Continental, Finger Food, Italian        1500.0   

                               listed_in(type)  
0                             Buffet, Dine-out  
1                           Delivery, Dine-out  
2  Dine-out, Drinks & nightlife, Pubs and bars  
3                                     Dine-out  
4  Dine-out, Drinks & nightlife, Pubs and bars

In [76]:
newDf.isnull().sum()

url                   0
name                  0
online_order          0
book_table            0
rating             1658
votes                 0
location              0
rest_type             0
cuisines              0
cost_for_two         55
listed_in(type)       0
dtype: int64

In [77]:
newDf[newDf.name=="Cafe Coffee Day"]

url             name  \
1914  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1915  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1916  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1917  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1918  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1919  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1920  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1921  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1922  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1923  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1924  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1925  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1926  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1927  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1928  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1929  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1930  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1931  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1932  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1933  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1934  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1935  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1936  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1937  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1938  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1939  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1940  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1941  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1942  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1943  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1944  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1945  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1946  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1947  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1948  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1949  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1950  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1951  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1952  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1953  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1954  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1955  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1956  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1957  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1958  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1959  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1960  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1961  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1962  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1963  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1964  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1965  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1966  https://www.zomato

In [78]:
df[df.name=="Cafe Coffee Day"]

url             name  \
22     https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
254    https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
625    https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
889    https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1186   https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1898   https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
1986   https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
2504   https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
2686   https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
3276   https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
3323   https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
3785   https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
4110   https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
4543   https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
4618   https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
5036   https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
5313   https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
5940   https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
5991   https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
6783   https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
6846   https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
7801   https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
8306   https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
8746   https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
10382  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
10501  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
12095  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
12504  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
13400  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
13544  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
14614  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
14911  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
15432  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
15828  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
16954  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
17078  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
17608  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
17844  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
19440  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
19879  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
20940  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
21151  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
21795  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
22120  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
23134  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
23265  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
23876  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
24050  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
24643  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
24740  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
25169  https://www.zomato.com/bangalore/cafe-coffee-d...  Cafe Coffee Day   
25356  https://www.zomato.com/bangalore/cafe-coff

#### **Filling rating null values after grouping**

In [79]:
# Group by restaurant name and check if any and not all ratings are null
rating_nulls = newDf.groupby('name')["rating"].agg(
    rating = lambda x : round(x.mean(),1),
    total_count=len,
    null_count=lambda x: x.isna().sum(),
    non_null_count=lambda x: x.notna().sum()
)

# Filter where there's a mix of null and non-null ratings
partial_nulls = rating_nulls.query("null_count > 0 and non_null_count > 0")

partial_nulls

rating  total_count  null_count  non_null_count
name                                                                      
7th Heaven                    3.6            2           1               1
9 Bowls                       0.0            4           3               1
AB's - Absolute Barbecues     4.8            6           1               5
Aahar Cafe                    3.0            4           2               2
Ambur Biriyani                3.4            3           1               2
...                           ...          ...         ...             ...
YumYumSouth                   3.4            6           1               5
Yumjar                        3.4            2           1               1
Z-Bar                         4.1            2           1               1
Zayakedar                     4.0            2           1               1
nu.tree                       4.3            4           1               3

[245 rows x 4 columns]

In [80]:
partial_nulls.null_count.sum()

399

*There are 399 null rate values of the same name of restaurant in some locations but not null values in other locations*\
*So we will fill these null values*

In [81]:
newDf["ratings"] = newDf.apply(lambda row : rating_nulls.loc[row["name"]].rating if pd.isna(row["rating"]) else row["rating"],axis=1)

In [82]:
newDf["ratings"].isnull().sum()

1259

*Now we have only 1259 null values left in ratings*

In [83]:
# Group by restaurant type and check if any and not all ratings are null
rating_nulls2 = newDf.groupby('rest_type')["ratings"].agg(
    ratings = lambda x : round(x.mean(),1),
    total_count=len,
    null_count=lambda x: x.isna().sum(),
    non_null_count=lambda x: x.notna().sum()
)

# Filter where there's a mix of null and non-null ratings
partial_nulls2 = rating_nulls2.query("null_count > 0 and non_null_count > 0")

partial_nulls2

ratings  total_count  null_count  \
rest_type                                                         
Bakery                             3.3          324          44   
Bakery, Cafe                       3.8           14           1   
Bakery, Dessert Parlor             3.5           35           3   
Bakery, Quick Bites                3.2           68           7   
Bar                                3.7           97           6   
Bar, Casual Dining                 3.7           58           1   
Beverage Shop                      3.1          254          18   
Beverage Shop, Dessert Parlor      3.8           30           1   
Beverage Shop, Quick Bites         3.6           91           4   
Cafe                               3.5          544          19   
Cafe, Bakery                       3.9           18           1   
Cafe, Quick Bites                  3.7           15           1   
Casual Dining                      3.6         2211          54   
Casual Dining, Bar                 3.8          152           1   
Confectionery                      3.9           14           3   
Delivery                           3.1          927         156   
Dessert Parlor                     3.6          559          29   
Dessert Parlor, Bakery             3.7           16           1   
Dessert Parlor, Beverage Shop      3.8           21           1   
Dessert Parlor, Quick Bites        3.8           25           2   
Dhaba                              2.4            8           1   
Fine Dining                        4.1           80           1   
Food Court                         3.2          186          14   
Food Truck                         3.6           28           3   
Kiosk                              3.4           62           2   
Mess                               3.4           76          14   
Pub, Microbrewery                  4.3            9           1   
Quick Bites                        3.2         5090         714   
Quick Bites, Bakery                3.7            8           1   
Quick Bites, Beverage Shop         3.7           28           7   
Quick Bites, Cafe                  3.2            9           1   
Quick Bites, Dessert Parlor        3.3           15           2   
Sweet Shop                         3.5          120          19   
Takeaway                           2.9           38           4   
Takeaway, Delivery                 2.9          750         117   

                               non_null_count  
rest_type                                      
Bakery                                    280  
Bakery, Cafe                               13  
Bakery, Dessert Parlor                     32  
Bakery, Quick Bites                        61  
Bar                                        91  
Bar, Casual Dining                         57  
Beverage Shop                             236  
Beverage Shop, Dessert Parlor              29  
Beverage Shop, Quick Bites                 87  
Cafe                                      525  
Cafe, Bakery                               17  
Cafe, Quick Bites                          14  
Casual Dining                            2157  
Casual Dining, Bar                        151  
Confectionery                              11  
Delivery                                  771  
Dessert Parlor                            530  
Dessert Parlor, Bakery                     15  
Dessert Parlor, Beverage Shop              20  
Dessert Parlor, Quick Bites                23  
Dhaba                                       7  
Fine Dining                                79  
Food Court                                172  
Food Truck                                 25  
Kiosk                                      60  
Mess                                       62  
Pub, Microbrewery                           8  
Quick Bites                              4376  
Quick Bites, Bakery                         7  
Quick Bites, Beverage Shop                 21  
Quick Bites, Ca

In [84]:
partial_nulls2.null_count.sum()

1254

*There are 1254 null rate values of the same restaurant type in some rows but not null values in others*\
*So we will fill these null values*

In [85]:
newDf["ratings"] = newDf.apply(lambda row : rating_nulls2.loc[row["rest_type"]].ratings if pd.isna(row["ratings"]) else row["ratings"],axis=1)

In [86]:
newDf["ratings"].isnull().sum()

5

*Still 5 null values left*\
*We will fill these with the overall means of ratings values*

In [87]:
newDf["ratings"] = newDf.ratings.fillna(newDf.ratings.mean())

In [88]:
newDf["ratings"] = newDf.ratings.map(lambda x : round(x,1))

In [89]:
newDf["ratings"].isnull().sum()

0

In [90]:
newDf.drop("rating",axis=1,inplace=True)

In [91]:
newDf = newDf.loc[:,newDf.columns[[0,1,2,3,10,4,5,6,7,8,9]]]

In [92]:
newDf.head()

url  \
0  https://www.zomato.com/24-7RestaurantBangalore   
1           https://www.zomato.com/29degreesnorth   
2           https://www.zomato.com/ArborBrewIndia   
3         https://www.zomato.com/BaluchiBangalore   
4                https://www.zomato.com/Barebones   

                                  name online_order book_table  ratings  \
0     24/7 - The Lalit Ashok Bangalore           No        Yes      4.0   
1                     29 Degrees North          Yes         No      3.9   
2                Arbor Brewing Company           No        Yes      4.5   
3  Baluchi - The Lalit Ashok Bangalore           No        Yes      4.0   
4          Barebones - The Balcony Bar           No        Yes      4.2   

   votes       location           rest_type  \
0    344  Seshadripuram         Fine Dining   
1    287      Banaswadi         Quick Bites   
2   8419   Brigade Road   Pub, Microbrewery   
3    134  Seshadripuram         Fine Dining   
4   2445    Indiranagar  Bar, Casual Dining   

                                      cuisines  cost_for_two  \
0  Continental, North Indian, Italian, Chinese        3000.0   
1         North Indian, Fast Food, Street Food         250.0   
2                        American, Continental        2000.0   
3              North Indian, Kashmiri, Mughlai        3700.0   
4   Chinese, Continental, Finger Food, Italian        1500.0   

                               listed_in(type)  
0                             Buffet, Dine-out  
1                           Delivery, Dine-out  
2  Dine-out, Drinks & nightlife, Pubs and bars  
3                                     Dine-out  
4  Dine-out, Drinks & nightlife, Pubs and bars

#### **Filling cost_for_two null values after grouping**

In [93]:
newDf.cost_for_two.isnull().sum()

55

In [94]:
# Group by restaurant type and check if any and not all ratings are null
rating_nulls3 = newDf.groupby('rest_type')["cost_for_two"].agg(
    cost_for_two = lambda x : round(x.mean(),-1),
    total_count=len,
    null_count=lambda x: x.isna().sum(),
    non_null_count=lambda x: x.notna().sum()
)

# Filter where there's a mix of null and non-null ratings
partial_nulls3 = rating_nulls3.query("null_count > 0 and non_null_count > 0")

partial_nulls3

cost_for_two  total_count  null_count  \
rest_type                                                           
Bakery, Quick Bites                380.0           68           1   
Bar                               1270.0           97           1   
Bar, Casual Dining                1320.0           58           1   
Beverage Shop, Quick Bites         260.0           91           1   
Cafe                               610.0          544           5   
Casual Dining                      760.0         2211          12   
Delivery                           440.0          927           7   
Dessert Parlor                     320.0          559           2   
Food Court                         440.0          186           2   
Quick Bites                        310.0         5090          19   
Takeaway                           310.0           38           1   
Takeaway, Delivery                 380.0          750           3   

                            non_null_count  
rest_type                                   
Bakery, Quick Bites                     67  
Bar                                     96  
Bar, Casual Dining                      57  
Beverage Shop, Quick Bites              90  
Cafe                                   539  
Casual Dining                         2199  
Delivery                               920  
Dessert Parlor                         557  
Food Court                             184  
Quick Bites                           5071  
Takeaway                                37  
Takeaway, Delivery                     747

In [95]:
partial_nulls3.null_count.sum()

55

*There are 55 null rate values of the same restaurant type in some rows but not null values in others*\
*So we will fill these null values*

In [96]:
newDf["cost_for_two"] = newDf.apply(lambda row : rating_nulls3.loc[row["rest_type"]].cost_for_two \
                                     if pd.isna(row["cost_for_two"]) else row["cost_for_two"],axis=1)

### **Final Cleaned Data**

In [97]:
newDf

url  \
0      https://www.zomato.com/24-7RestaurantBangalore   
1               https://www.zomato.com/29degreesnorth   
2               https://www.zomato.com/ArborBrewIndia   
3             https://www.zomato.com/BaluchiBangalore   
4                    https://www.zomato.com/Barebones   
...                                               ...   
12448          https://www.zomato.com/thesugarfactory   
12449          https://www.zomato.com/theunderpassblr   
12450               https://www.zomato.com/ucfastfood   
12451             https://www.zomato.com/umashomefood   
12452        https://www.zomato.com/zyksha-food-truck   

                                      name online_order book_table  ratings  \
0         24/7 - The Lalit Ashok Bangalore           No        Yes      4.0   
1                         29 Degrees North          Yes         No      3.9   
2                    Arbor Brewing Company           No        Yes      4.5   
3      Baluchi - The Lalit Ashok Bangalore           No        Yes      4.0   
4              Barebones - The Balcony Bar           No        Yes      4.2   
...                                    ...          ...        ...      ...   
12448      The Sugar Factory - Le Meridien           No         No      3.9   
12449             The Underpass Pub & Grub           No        Yes      4.0   
12450               United Cafe & Fastfood           No         No      3.2   
12451                      Uma's Home Food           No         No      3.6   
12452                               Zyksha           No         No      3.4   

       votes         location           rest_type  \
0        344    Seshadripuram         Fine Dining   
1        287        Banaswadi         Quick Bites   
2       8419     Brigade Road   Pub, Microbrewery   
3        134    Seshadripuram         Fine Dining   
4       2445      Indiranagar  Bar, Casual Dining   
...      ...              ...                 ...   
12448    630      Sankey Road                Club   
12449    726         JP Nagar  Pub, Casual Dining   
12450      0     New BEL Road         Quick Bites   
12451     15     Marathahalli            Delivery   
12452      9  South Bangalore          Food Truck   

                                              cuisines  cost_for_two  \
0          Continental, North Indian, Italian, Chinese        3000.0   
1                 North Indian, Fast Food, Street Food         250.0   
2                                American, Continental        2000.0   
3                      North Indian, Kashmiri, Mughlai        3700.0   
4           Chinese, Continental, Finger Food, Italian        1500.0   
...                                                ...           ...   
12448                                      Finger Food        2000.0   
12449  Finger Food, Continental, North Indian, Chinese        1500.0   
12450                          Fast Food, North Indian         400.0   
12451                                     Healthy Food         300.0   
12452                                        Fast Food         200.0   

                                   listed_in(type)  
0                                 Buffet, Dine-out  
1                               Delivery, Dine-out  
2      Dine-out, Drinks & nightlife, Pubs and bars  
3                                         Dine-out  
4      Dine-out, Drinks & nightlife, Pubs and bars  
...                                            ...  
12448                 Dine-out, Drinks & nightlife  
12449                 Dine-out, Drinks & nightlife  
12450                           Delivery, Dine-out  
12451                                     Delivery  
12452                           Delivery, Dine-out  

[12453 rows x 11 columns]

In [98]:
newDf.isnull().sum()

url                0
name               0
online_order       0
book_table         0
ratings            0
votes              0
location           0
rest_type          0
cuisines           0
cost_for_two       0
listed_in(type)    0
dtype: int64

In [99]:
newDf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12453 entries, 0 to 12452
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   url              12453 non-null  object 
 1   name             12453 non-null  object 
 2   online_order     12453 non-null  object 
 3   book_table       12453 non-null  object 
 4   ratings          12453 non-null  float64
 5   votes            12453 non-null  int64  
 6   location         12453 non-null  object 
 7   rest_type        12453 non-null  object 
 8   cuisines         12453 non-null  object 
 9   cost_for_two     12453 non-null  float64
 10  listed_in(type)  12453 non-null  object 
dtypes: float64(2), int64(1), object(8)
memory usage: 1.0+ MB


**There are no null values and all columns have their appropriate data types**\
**So we can consider our data has been cleaned successfully**

In [100]:
del df

In [101]:
df=newDf

In [102]:
df.to_csv("Fully Cleaned Zomato Data.csv",index=False)